nnc/Store.swift

```diff
@@ -147,71 +200,114 @@ private let fpzipDecode:
+      let decodedSize = decodedSize, dimensionCount > 0
+    else { return 0 }
+    var floatCount = decodedSize[0] / MemoryLayout<Float16>.size
+    let exponentZipSize = Int(data.assumingMemoryBound(to: Int32.self)[0])
+    let exponentZipData = data.advanced(by: MemoryLayout<Int32>.size)
+    let exponentBuffer = UnsafeMutablePointer<UInt8>.allocate(capacity: floatCount)
+    defer { exponentBuffer.deallocate() }
+    guard unzip(data: exponentZipData,
+                dataSize: exponentZipSize,
+                unzippedData: exponentBuffer,
+                unzippedDataSize: &floatCount) else { return 0 }
+    let floatsWithoutExp = exponentZipData.advanced(by: exponentZipSize).assumingMemoryBound(to: UInt8.self)
+    let decodedAsInts = decoded.assumingMemoryBound(to: UInt16.self)
+    for i in 0..<floatCount {
+      let floatWithoutExp = UInt16(floatsWithoutExp[i])
+      let signBit = floatWithoutExp >> 7
+      let mantissa = (floatWithoutExp & 0x7f) << 3
+      let exponent = UInt16(exponentBuffer[i])
+      decodedAsInts[i] = (signBit << 15) | (exponent << 10) | mantissa
     }
+
+    return 1
+  }
+
+#if canImport(Foundation) && canImport(Compression)
+  import Foundation
+  import Compression
+
+  func zip(data: UnsafeRawPointer, dataSize: Int, zippedData: UnsafeMutableRawPointer, zippedDataSize: UnsafeMutablePointer<Int>) -> Bool {
+    let outputSize = compression_encode_buffer(
+      zippedData.assumingMemoryBound(to: UInt8.self), zippedDataSize[0],
+      data.assumingMemoryBound(to: UInt8.self), dataSize, nil, COMPRESSION_ZLIB)
+    guard outputSize > 0 else { return false }
+    zippedDataSize[0] = outputSize
+    return true
+  }
+
+  private func unzip(data: UnsafeRawPointer, dataSize: Int, unzippedData: UnsafeMutableRawPointer, unzippedDataSize: UnsafeMutablePointer<Int>) -> Bool {
+    let data = NSData(bytesNoCopy: UnsafeMutableRawPointer(mutating: data), length: dataSize, deallocator: .none)
+    guard let decompressedData = try? data.decompressed(using: .zlib) else { return false }
+    guard decompressedData.length <= unzippedDataSize[0] else { return false }
+    unzippedDataSize[0] = decompressedData.length
+    memcpy(unzippedData, decompressedData.bytes, decompressedData.length)
+    return true
+  }
+
 #else
-  private let zipEncode:
-    @convention(c) (
-      UnsafeRawPointer?, Int, Int32, UnsafePointer<Int32>?, Int32, UnsafeMutableRawPointer?,
-      UnsafeMutableRawPointer?, UnsafeMutablePointer<Int>?, UnsafeMutablePointer<UInt32>?
-    ) -> Int32 = {
-      data, dataSize, dataType, dimensions, dimensionCount, context, encoded, encodedSize,
-      identifier
-      in
-      guard let data = data, let dimensions = dimensions, let encoded = encoded,
-        let encodedSize = encodedSize, dimensionCount > 0
-      else { return 0 }
+
+  private func zip(data: UnsafeRawPointer, dataSize: Int, zippedData: UnsafeMutablePointer<Int>, zippedDataSize: UnsafeMutablePointer<Int>) -> Bool {
```

Comment:
I am actually unsure about this mapping. I thought Swift's Bool maps to Int8?

Reply:
What are you concerned with exactly? This function is only called by other swift functions, so it doesn't need to worry about C interop

Reply:
Oh, nvm then, I missed that part.